sycl/include/CL/sycl/detail/image_accessor_util.hpp

```diff
@@ -123,85 +188,528 @@ template <typename T> vec<T, 4> readPixel(T *Ptr, image_channel_order Order) {
+    throw cl::sycl::invalid_parameter_error("Unhandled image channel order");
     break;
   }
 }
 
 // Converts read pixel data into return datatype based on the channel type of
 // the image.
-// TODO: Change this method to use the conversion rules as given in the OpenCL
+// Conversion rules used as given in the OpenCL
 // Spec section 8.3. The conversion rules may be handled differently for each
 // return datatype - float, int32, uint32, half. ImageChannelType is passed to
 // the function to use appropriate conversion rules.
-template <typename ChannelType, typename RetDataType,
-          typename = detail::enable_if_t<
-              (detail::is_contained<
-                  RetDataType, type_list<cl_int, cl_float, cl_uint>>::value)>>
+
+template <typename ChannelType>
+void convertReadData(vec<ChannelType, 4> PixelData,
+                     const image_channel_type ImageChannelType,
+                     vec<cl_uint, 4> &RetData) {
+
+  switch (ImageChannelType) {
+  case image_channel_type::unsigned_int8:
+  case image_channel_type::unsigned_int16:
+  case image_channel_type::unsigned_int32:
+    RetData = PixelData.template convert<cl_uint>();
+    break;
+  default:
+    // OpenCL Spec section 6.12.14.2 does not allow reading uint4 data from an
+    // image with channel datatype other than unsigned_int8,unsigned_int16 and
+    // unsigned_int32.
+    throw cl::sycl::invalid_parameter_error(
+        "Datatype of read data - cl_uint4 is incompatible with the "
+        "image_channel_type of the image.");
+    break;
+  }
+}
+
+template <typename ChannelType>
+void convertReadData(vec<ChannelType, 4> PixelData,
+                     image_channel_type ImageChannelType,
+                     vec<cl_int, 4> &RetData) {
+
+  switch (ImageChannelType) {
+  case image_channel_type::signed_int8:
+  case image_channel_type::signed_int16:
+  case image_channel_type::signed_int32:
+    RetData = PixelData.template convert<cl_int>();
+    break;
+  default:
+    // OpenCL Spec section 6.12.14.2 does not allow reading int4 data from an
+    // image with channel datatype other than signed_int8,signed_int16 and
+    // signed_int32.
+    throw cl::sycl::invalid_parameter_error(
+        "Datatype of read data - cl_int4 is incompatible with the "
+        "image_channel_type of the image.");
+    break;
+  }
+}
+
+template <typename ChannelType>
 void convertReadData(vec<ChannelType, 4> PixelData,
                      image_channel_type ImageChannelType,
-                     vec<RetDataType, 4> &RetData) {
-  RetData.x() = (RetDataType)PixelData.x();
-  RetData.y() = (RetDataType)PixelData.y();
-  RetData.z() = (RetDataType)PixelData.z();
-  RetData.w() = (RetDataType)PixelData.w();
+                     vec<cl_float, 4> &RetData) {
+
+  vec<cl_float, 4> Temp;
+
+  switch (ImageChannelType) {
+  case image_channel_type::snorm_int8:
+    //  max(-1.0f, (float)c / 127.0f)
+    RetData.x() = (cl_float)PixelData.x() / 127.0f;
```

Comment:
Perhaps you could replace everywhere all the old C-style casts by some `static_cast<cl_float>(PixelData.x())` or `cl_float { PixelData.x() }` according to the various data types